File: Sources/Sentry/SentryWatchdogTerminationScopeObserver.m
#import "SentryWatchdogTerminationScopeObserver.h"

#if SENTRY_HAS_UIKIT

#    import <SentryBreadcrumb.h>
#    import <SentryFileManager.h>
#    import <SentryLogC.h>
#    import <SentrySwift.h>
#    import <SentryWatchdogTerminationBreadcrumbProcessor.h>

@interface SentryWatchdogTerminationScopeObserver ()

@property (nonatomic, strong) SentryWatchdogTerminationBreadcrumbProcessor *breadcrumbProcessor;
@property (nonatomic, strong) SentryWatchdogTerminationAttributesProcessor *attributesProcessor;

@end

@implementation SentryWatchdogTerminationScopeObserver

- (instancetype)
    initWithBreadcrumbProcessor:(SentryWatchdogTerminationBreadcrumbProcessor *)breadcrumbProcessor
            attributesProcessor:(SentryWatchdogTerminationAttributesProcessor *)attributesProcessor;
{
    if (self = [super init]) {
        self.breadcrumbProcessor = breadcrumbProcessor;
        self.attributesProcessor = attributesProcessor;
    }

    return self;
}

// PRAGMA MARK: - SentryScopeObserver

- (void)clear
{
    [self.breadcrumbProcessor clear];
    [self.attributesProcessor clear];
}

- (void)addSerializedBreadcrumb:(NSDictionary *)crumb
{
    [self.breadcrumbProcessor addSerializedBreadcrumb:crumb];
}

- (void)clearBreadcrumbs
{
    [self.breadcrumbProcessor clearBreadcrumbs];
}

- (void)setContext:(nullable NSDictionary<NSString *, id> *)context
{
    [self.attributesProcessor setContext:context];
}

- (void)setDist:(nullable NSString *)dist
{
    // Left blank on purpose
}

- (void)setEnvironment:(nullable NSString *)environment
{
    // Left blank on purpose
}

- (void)setExtras:(nullable NSDictionary<NSString *, id> *)extras
{
    // Left blank on purpose
}

- (void)setFingerprint:(nullable NSArray<NSString *> *)fingerprint
{
    // Left blank on purpose
}

- (void)setLevel:(enum SentryLevel)level
{
    // Left blank on purpose
}

- (void)setTags:(nullable NSDictionary<NSString *, NSString *> *)tags
{
    // Left blank on purpose
}

- (void)setUser:(nullable SentryUser *)user
{
    [self.attributesProcessor setUser:user];
}

- (void)setTraceContext:(nullable NSDictionary<NSString *, id> *)traceContext
{
    // Left blank on purpose
}

@end

#endif // SENTRY_HAS_UIKIT

File: Sources/Swift/Integrations/Session/SentryWatchdogTerminationScopeObserver.swift
@_implementationOnly import _SentryPrivate

/**
 * This scope observer is used by the Watchdog Termination integration to write breadcrumbs to disk.
 * The overhead is ~0.015 seconds for 1000 breadcrumbs.
 * This class doesn't need to be thread safe as the scope already calls the scope observers in a
 * thread safe manner.
 */
@_spi(Private) @objc public class SentryWatchdogTerminationScopeObserver: NSObject, SentryScopeObserver {
    
    private let breadcrumbProcessor: SentryWatchdogTerminationBreadcrumbProcessor
    private let attributesProcessor: SentryWatchdogTerminationAttributesProcessor
    
    init(breadcrumbProcessor: SentryWatchdogTerminationBreadcrumbProcessor, attributesProcessor: SentryWatchdogTerminationAttributesProcessor) {
        self.breadcrumbProcessor = breadcrumbProcessor
        self.attributesProcessor = attributesProcessor
    }

    @objc public convenience init(maxBreadcrumbs: Int, attributesProcessor: SentryWatchdogTerminationAttributesProcessor) {
        self.init(breadcrumbProcessor: SentryWatchdogTerminationBreadcrumbProcessor(maxBreadcrumbs: maxBreadcrumbs), attributesProcessor: attributesProcessor)
    }

    public func setUser(_ user: User?) {
        attributesProcessor.setUser(user)
    }
    
    public func setTags(_ tags: [String: String]?) {
        attributesProcessor.setTags(tags)
    }
    
    public func setExtras(_ extras: [String: Any]?) {
        attributesProcessor.setExtras(extras)
    }
    
    public func setContext(_ context: [String: [String: Any]]?) {
        attributesProcessor.setContext(context)
    }
    
    public func setTraceContext(_ traceContext: [String: Any]?) {
        // Nothing to do here, Trace Context is not persisted for watchdog termination events
        // On regular events, we have the current trace in memory, but there isn't time to persist one
        // in watchdog termination events
    }
    
    public func setDist(_ dist: String?) {
        attributesProcessor.setDist(dist)
    }
    
    public func setEnvironment(_ environment: String?) {
        attributesProcessor.setEnvironment(environment)
    }
    
    public func setFingerprint(_ fingerprint: [String]?) {
        attributesProcessor.setFingerprint(fingerprint)
    }
    
    public func setLevel(_ level: SentryLevel) {
        // Nothing to do here, watchdog termination events are always Fatal
    }
    
    public func addSerializedBreadcrumb(_ serializedBreadcrumb: [String: Any]) {
        breadcrumbProcessor.addSerializedBreadcrumb(serializedBreadcrumb)
    }
    
    public func clearBreadcrumbs() {
        breadcrumbProcessor.clearBreadcrumbs()
    }
    
    public func clear() {
        breadcrumbProcessor.clear()
        attributesProcessor.clear()
    }
}
